src/pdmfilter.rs

```diff
@@ -0,0 +1,196 @@
+    lut: [u32; (SINCN * PDM_DECIMATION / 8) as usize * 256],
+    fs: u32,
+    coef: [u32; 3],
+    sub_const: u32,
+    old_out: i64,
+    old_in: i64,
+    oldz: i64,
+    hp_alpha: u32,
+    lp_alpha: u32,
+    div_const: u32,
+}
+impl PDMFilter {
+    pub fn new(sample_rate: u32) -> PDMFilter {
+        PDMFilter {
+            lut: [0u32; (SINCN * PDM_DECIMATION / 8) as usize * 256],
+            fs: sample_rate * PDM_DECIMATION as u32,
+            coef: [0, 0, 0],
+            sub_const: 0,
+            old_out: 0,
+            old_in: 0,
+            oldz: 0,
+            lp_alpha: 0,
+            hp_alpha: 0,
+            div_const: 0,
+        }
+    }
+    pub fn init(&mut self) {
+        let lp_hz: f32 = self.fs as f32 / 128.0;
+        let hp_hz: f32 = 10.0;
+        if lp_hz != 0.0 {
+            self.lp_alpha = (lp_hz as f32 * 256.0 / (lp_hz + self.fs as f32 / (2.0 * PI))) as u32;
+        }
+        if hp_hz != 0.0 {
+            self.hp_alpha = (self.fs as f32 * 256.0 / (2.0 * PI * hp_hz + self.fs as f32)) as u32;
+        }
+        let mut sinc = [1u16; PDM_DECIMATION as usize];
+        let mut sinc_out = [0u16; PDM_DECIMATION as usize * 2 - 1];
+
+        let mut sinc2 = [0u16; PDM_DECIMATION as usize * 3 - 2];
+        convolve(
+            &sinc,
+            PDM_DECIMATION as usize,
+            &sinc,
+            PDM_DECIMATION as usize,
+            sinc_out.as_mut(),
+        );
+        convolve(
+            &sinc_out,
+            PDM_DECIMATION as usize * 2 - 1,
+            &sinc,
+            PDM_DECIMATION as usize,
+            sinc2.as_mut(),
+        );
+        let mut sum: u32 = 0;
+        let mut coef = [0u32; ((PDM_DECIMATION * SINCN) as usize + 2)];
+        for j in 0..SINCN {
+            for i in 0..PDM_DECIMATION {
+                // seems like this shouldn't happen...
+                if (j * PDM_DECIMATION + i) as usize >= sinc2.len() {
+                    continue;
+                    // coef[(j * PDM_DECIMATION + i) as usize] = 0;
+                } else {
+                    coef[(j * PDM_DECIMATION + i + 1) as usize] =
+                        sinc2[(j * PDM_DECIMATION + i) as usize] as u32;
+                    sum += sinc2[(j * PDM_DECIMATION + i) as usize] as u32;
+                }
+                // if j >= 1 {
+                // break;?
+                // }
+            }
+        }
+        self.sub_const = sum >> 1;
+        self.div_const = self.sub_const * MAX_VOLUME as u32 / 32768 / FILTER_GAIN as u32;
+        if self.div_const == 0 {
+            self.div_const = 1;
+        }
+
+        for s in 0..SINCN {
+            let offset: usize = (s * PDM_DECIMATION) as usize;
+
+            for c in 0..256u32 {
+                for d in 0..8 as usize {
+                    self.lut[((s as usize * 256 * 8) + c as usize * 8 + d) as usize] = (c >> 7)
+                        * coef[offset + d * 8]
+                        + ((c >> 6) & 0x01) * coef[offset + d * 8 + 1]
```

Comment:
I'm guessing the compiler is probably smart enough to simplify the common parts of these index expressions, but if we're unsure we could always make `offset` const, and also assign `offset + d * 8` to an intermediate variable?  Probably doesn't matter though, other than making it more readable.